plugins/modules/grafana_folder.py

```diff
@@ -0,0 +1,277 @@
+        # {{{ Authentication header
+        self.headers = {"Content-Type": "application/json"}
+        if module.params.get('grafana_api_key', None):
+            self.headers["Authorization"] = "Bearer %s" % module.params['grafana_api_key']
+        else:
+            self.headers["Authorization"] = basic_auth_header(module.params['url_username'], module.params['url_password'])
+        # }}}
+        self.grafana_url = module.params.get("url")
+        grafana_version = self.get_version()
+        if grafana_version["major"] < 5:
+            self._module.fail_json(failed=True, msg="Folders API is available starting Grafana v5")
+
+    def _send_request(self, url, data=None, headers=None, method="GET"):
+        if data is not None:
+            data = json.dumps(data, sort_keys=True)
+        if not headers:
+            headers = []
+
+        full_url = "{grafana_url}{path}".format(grafana_url=self.grafana_url, path=url)
+        resp, info = fetch_url(self._module, full_url, data=data, headers=headers, method=method)
+        status_code = info["status"]
+        if status_code == 404:
+            return None
+        elif status_code == 401:
+            self._module.fail_json(failed=True, msg="Unauthorized to perform action '%s' on '%s'" % (method, full_url))
+        elif status_code == 403:
+            self._module.fail_json(failed=True, msg="Permission Denied")
+        elif status_code == 412:
+            error_msg = resp.read()['message']
+            self._module.fail_json(failed=True, msg=error_msg)
+        elif status_code == 200:
+            return self._module.from_json(resp.read())
+        self._module.fail_json(failed=True, msg="Grafana Folders API answered with HTTP %d" % status_code)
+
+    def get_version(self):
+        url = "/api/health"
+        response = self._send_request(url, data=None, headers=self.headers, method="GET")
+        version = response.get("version")
+        major, minor, rev = version.split(".")
+        return {"major": int(major), "minor": int(minor), "rev": int(rev)}
+
+    def create_folder(self, title):
+        url = "/api/folders"
+        folder = dict(title=title)
+        response = self._send_request(url, data=folder, headers=self.headers, method="POST")
+        return response
+
+    def get_folder(self, title):
+        url = "/api/folders?limit=1000"
+        response = self._send_request(url, headers=self.headers, method="GET")
+        for item in response:
+            if item.get("title") == title:
+                return item
+        return None
+
+    def delete_folder(self, folder_uid):
+        url = "/api/folders/{folder_uid}".format(folder_uid=folder_uid)
+        response = self._send_request(url, headers=self.headers, method="DELETE")
+        return response
+
+
+def setup_module_object():
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        supports_check_mode=False,
+        required_together=[['url_username', 'url_password']],
```

Comment:
`required_together` and `mutually_exclusive` should be moved to `plugins/module_utils/base.py` and then removed from the individual modules.

Reply:
I created 2 functions in `base.py` and called them in the "new" modules introduced in the collection.